plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -1155,50 +1159,90 @@ public void renameTable(ConnectorSession session, ConnectorTableHandle tableHand
-                    updateProperty(updateProperties, FORMAT_VERSION, propertyValue, formatVersion -> Integer.toString((int) formatVersion));
+                    updateProperties = updateProperties.or(() -> Optional.of(transaction.updateProperties()))
+                            .map(update -> updateProperty(update, FORMAT_VERSION, propertyValue, formatVersion -> Integer.toString((int) formatVersion)));
+                    break;
+                case PARTITIONING_PROPERTY:
+                    // Partition updating is handled after the UpdateProperties is committed
                     break;
                 default:
-                    // TODO: Support updating partitioning https://github.com/trinodb/trino/issues/12174
                     throw new TrinoException(NOT_SUPPORTED, "Updating the " + trinoPropertyName + " property is not supported");
             }
         }
 
         try {
-            updateProperties.commit();
+            updateProperties.ifPresent(PendingUpdate::commit);
+            if (properties.containsKey(PARTITIONING_PROPERTY)) {
+                updatePartitioning(icebergTable, transaction, (List<String>) properties.get(PARTITIONING_PROPERTY).orElseThrow());
+            }
             transaction.commitTransaction();
         }
         catch (RuntimeException e) {
             throw new TrinoException(ICEBERG_COMMIT_ERROR, "Failed to commit new table properties", e);
         }
     }
 
-    private static void updateProperty(UpdateProperties updateProperties, String icebergPropertyName, Optional<Object> value, Function<Object, String> toIcebergString)
+    private static UpdateProperties updateProperty(UpdateProperties updateProperties, String icebergPropertyName, Optional<Object> value, Function<Object, String> toIcebergString)
     {
         if (value.isPresent()) {
             updateProperties.set(icebergPropertyName, toIcebergString.apply(value.get()));
         }
         else {
             updateProperties.remove(icebergPropertyName);
         }
+
+        return updateProperties;
+    }
+
+    private static void updatePartitioning(Table icebergTable, Transaction transaction, List<String> partitionColumns)
+    {
+        UpdatePartitionSpec updatePartitionSpec = transaction.updateSpec();
+        Set<PartitionField> existingPartitionFields = icebergTable.spec().fields().stream().collect(toImmutableSet());
+        Schema schema = icebergTable.schema();
+        if (partitionColumns.isEmpty()) {
+            existingPartitionFields.stream()
+                    .map(partitionField -> toIcebergTerm(schema, partitionField))
+                    .forEach(updatePartitionSpec::removeField);
+        }
+        else {
+            List<PartitionField> partitionFields = parsePartitionFields(schema, partitionColumns).fields();
+            existingPartitionFields.stream()
+                    .filter(partitionField -> !partitionFields.contains(partitionField))
```

Comment:
nit: quadratic

make `partitionFields` a `Set` (`ImmutableSet` maintains ordering)
and use `Sets.difference`  to symmetrically iterate over fields to remove and fields to add